zigpy/device.py

```diff
@@ -47,118 +49,191 @@ class Device(zigpy.util.LocalLogMixin, zigpy.util.ListenableMixin):
+        return self._group_scan_task
 
     async def group_membership_scan(self) -> None:
         """Sync up group membership."""
         for ep_id, ep in self.endpoints.items():
             if ep_id:
                 await ep.group_membership_scan()
 
-    async def get_node_descriptor(self):
+    @property
+    def initializing(self) -> bool:
+        """Return True if device is being initialized."""
+        return self._initialize_task is not None and not self._initialize_task.done()
+
+    def cancel_initialization(self) -> None:
+        """Cancel initialization call."""
+        if self.initializing:
+            self.debug("Canceling old initialize call")
+            self._initialize_task.cancel()
+
+    def schedule_initialize(self) -> Optional[asyncio.Task]:
+        # Already-initialized devices don't need to be re-initialized
+        if self.is_initialized:
+            self.debug("Skipping initialization, device is fully initialized")
+            self._application.device_initialized(self)
+            return None
+
+        self.debug("Scheduling initialization")
+
+        self.cancel_initialization()
+        self._initialize_task = asyncio.create_task(self.initialize())
+
+        return self._initialize_task
+
+    async def get_node_descriptor(self) -> zdo.types.NodeDescriptor:
         self.info("Requesting 'Node Descriptor'")
-        try:
-            status, _, node_desc = await self.zdo.Node_Desc_req(
-                self.nwk, tries=2, delay=0.1
+
+        status, _, node_desc = await self.zdo.Node_Desc_req(
+            self.nwk, tries=2, delay=0.1
+        )
+
+        if status != zdo.types.Status.SUCCESS:
+            raise zigpy.exceptions.InvalidResponse(
+                f"Requesting Node Descriptor failed: {status}"
             )
-            if status == zdo.types.Status.SUCCESS:
-                self.node_desc = node_desc
-                self.info("Node Descriptor: %s", node_desc)
-                return node_desc
-            else:
-                self.warning("Requesting Node Descriptor failed: %s", status)
-        except (asyncio.TimeoutError, zigpy.exceptions.ZigbeeException):
-            self.warning("Requesting Node Descriptor failed", exc_info=True)
+
+        self.node_desc = node_desc
+        self.info("Got Node Descriptor: %s", node_desc)
+
+        return node_desc
+
+    async def initialize(self, *, tries=3, delay=0.5):
+        """
+        Attempts multiple times to discover all basic information about a device: namely
+        its node descriptor, all endpoints and clusters, and the model and manufacturer
+        attributes from any Basic cluster exposing those attributes.
+        """
+
+        for attempt in range(1, tries + 1):
+            try:
+                await self._initialize()
+                break
+            except asyncio.CancelledError:
+                raise
+            except Exception:
```

Comment:
Is this too "broad" ? Should retry if it is a timeout or a ZigBee exception, for any other reason it is a BUG and should throw the exception